File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> ans;
        for(int i=0; i<(int)nums.size(); ++i){
            if(ans.empty() || ans.back()<nums[i]) ans.push_back(nums[i]);
            else{
                int idx=lower_bound(ans.begin(), ans.end(), nums[i]) - ans.begin();
                ans[idx]= nums[i];
            }
        }
        return ans.size();
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n= nums.length, idx=0;
        int[] tmp= new int[n];
        tmp[0]= nums[0];
        for(int i=1;i<n;i++){
            if(nums[i]>tmp[idx]){
                tmp[++idx]= nums[i];
                continue;
            }
            int lo=0, hi= idx, mid, ans=0;
            while(lo<=hi){
                mid= (lo+hi)/2;
                if(tmp[mid]==nums[i]){
                    ans= mid; break;
                }
                else if(tmp[mid]>nums[i]){
                    ans= mid; hi= mid-1;
                }
                else{
                    lo= mid+1;
                }
            }
            tmp[ans]= nums[i];
        }
        return idx+1;
    }
}